test/unittest/memory/memory_planner_validate.cpp

```diff
@@ -0,0 +1,173 @@
+
+  return true;
+}
+
+/**
+ * @brief Validate the provided layout does completely overlap
+ */
+static bool validateAllOverlap(const std::vector<size_t> &memory_size,
+                               const std::vector<size_t> &memory_offset) {
+  for (unsigned int idx = 0; idx < memory_size.size(); idx++) {
+    EXPECT_EQ(memory_offset[idx], memory_offset[0]);
+  }
+
+  return true;
+}
+
+/**
+ * @brief Validate the provided layout does not overflow outside the given
+ * size
+ */
+static bool validateOverflow(const std::vector<size_t> &memory_size,
+                             const std::vector<size_t> &memory_offset,
+                             size_t pool_size) {
+  for (unsigned int idx = 0; idx < memory_size.size(); idx++)
+    if (memory_offset[idx] + memory_size[idx] > pool_size)
+      return false;
+
+  return true;
+}
+
+/**
+ * @brief Test memory planning with full overlap of requested memories
+ */
+TEST_P(MemoryPlannerValidate, full_overlap) {
+  std::mt19937 rng;
+  std::uniform_int_distribution<size_t> dist(1, MEM_BYTES);
+
+  std::vector<size_t> memory_size(MEM_QUANT);
+  for (unsigned int idx = 0; idx < memory_size.size(); idx++) {
+    memory_size[idx] = dist(rng);
+  }
+  std::vector<std::pair<unsigned int, unsigned int>> memory_validity(MEM_QUANT,
+                                                                     {1, 2});
+  std::vector<size_t> memory_offset;
+
+  size_t pool_size =
+    planner->planLayout(memory_size, memory_validity, memory_offset);
+
+  EXPECT_EQ(pool_size,
+            std::accumulate(memory_size.begin(), memory_size.end(), 0));
+  EXPECT_TRUE(validateNoOverlap(memory_size, memory_offset, pool_size));
+  EXPECT_TRUE(validateOverflow(memory_size, memory_offset, pool_size));
+}
+
+/**
+ * @brief Test memory planning with no overlap of requested memories
+ */
+TEST_P(MemoryPlannerValidate, none_overlap) {
+  std::mt19937 rng;
+  std::uniform_int_distribution<size_t> dist(1, MEM_BYTES);
+
+  std::vector<size_t> memory_size(MEM_QUANT);
+  for (unsigned int idx = 0; idx < memory_size.size(); idx++) {
+    memory_size[idx] = dist(rng);
+  }
+
+  std::vector<std::pair<unsigned int, unsigned int>> memory_validity(MEM_QUANT);
+  for (unsigned int idx = 0; idx < memory_validity.size(); idx++) {
+    memory_validity[idx] = {idx, idx + 1};
+  }
+
+  std::vector<size_t> memory_offset;
+  size_t pool_size =
+    planner->planLayout(memory_size, memory_validity, memory_offset);
+
+  EXPECT_TRUE(validateOverflow(memory_size, memory_offset, pool_size));
+  if (planner->getType() == nntrainer::BasicPlanner::type) {
+    EXPECT_EQ(pool_size,
+              std::accumulate(memory_size.begin(), memory_size.end(), 0));
+    EXPECT_TRUE(validateNoOverlap(memory_size, memory_offset, pool_size));
+  } else {
+    EXPECT_EQ(pool_size,
+              *std::max_element(memory_size.begin(), memory_size.end()));
+    EXPECT_TRUE(validateAllOverlap(memory_size, memory_offset));
```

Comment:
I wonder what this is for, is this checking planner->planLayerout is noop if nothing has been done?

Reply:
As there is no overlap, a good planner should give share a common memory to all the requests. 
The check is very rigid in the way that it expects all the memories to have same starting point. The check will be updated once other better planners are implemented.

Reply:
Yup